File: ios/MenuMakerUITests/PageObjects/OrderHistoryPage.swift
//
//  OrderHistoryPage.swift
//  MenuMakerUITests
//
//  Page Object for Order History Screen
//

import XCTest

struct OrderHistoryPage {
    let app: XCUIApplication

    // MARK: - Elements

    var ordersList: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "OrderItem")
    }

    var firstOrder: XCUIElement {
        ordersList.firstMatch
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no order'")).firstMatch
    }

    var activeTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'active' OR label CONTAINS[c] 'ongoing'")).firstMatch
    }

    var completedTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'completed' OR label CONTAINS[c] 'past'")).firstMatch
    }

    var cancelledTab: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'cancelled'")).firstMatch
    }

    var filterButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'filter'")).firstMatch
    }

    var searchBar: XCUIElement {
        app.searchFields.firstMatch
    }

    var reorderButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'reorder' OR label CONTAINS[c] 'order again'"))
    }

    var trackButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'track'"))
    }

    var viewDetailsButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'detail' OR label CONTAINS[c] 'view'"))
    }

    var orderIdLabels: XCUIElementQuery {
        app.staticTexts.matching(NSPredicate(format: "label MATCHES '.*#[A-Z0-9]{6,12}.*'"))
    }

    var orderDateLabels: XCUIElementQuery {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'jan' OR label CONTAINS[c] 'feb' OR label CONTAINS[c] 'mar' OR label CONTAINS[c] 'apr' OR label CONTAINS[c] 'may' OR label CONTAINS[c] 'jun' OR label CONTAINS[c] 'today' OR label CONTAINS[c] 'yesterday'"))
    }

    var orderTotalLabels: XCUIElementQuery {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS '₹'"))
    }

    var helpButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'help' OR label CONTAINS[c] 'support'")).firstMatch
    }

    // Date range filter elements
    var dateRangeButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'date' OR label CONTAINS[c] 'range'")).firstMatch
    }

    var last7DaysButton: XCUIElement {
        app.buttons["Last 7 Days"]
    }

    var last30DaysButton: XCUIElement {
        app.buttons["Last 30 Days"]
    }

    var last3MonthsButton: XCUIElement {
        app.buttons["Last 3 Months"]
    }

    // MARK: - Actions

    @discardableResult
    func tapFirstOrder() -> OrderHistoryPage {
        firstOrder.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func tapOrder(at index: Int) -> OrderHistoryPage {
        let order = ordersList.element(boundBy: index)
        if order.waitForExistence(timeout: 2) {
            order.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToActiveOrders() -> OrderHistoryPage {
        if activeTab.waitForExistence(timeout: 2) {
            activeTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToCompletedOrders() -> OrderHistoryPage {
        if completedTab.waitForExistence(timeout: 2) {
            completedTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func switchToCancelledOrders() -> OrderHistoryPage {
        if cancelledTab.waitForExistence(timeout: 2) {
            cancelledTab.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func reorderFirst() -> OrderHistoryPage {
        let reorderButton = reorderButtons.firstMatch
        if reorderButton.waitForExistence(timeout: 2) {
            reorderButton.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func trackFirstOrder() -> OrderHistoryPage {
        let trackButton = trackButtons.firstMatch
        if trackButton.waitForExistence(timeout: 2) {
            trackButton.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func searchOrders(_ query: String) -> OrderHistoryPage {
        if searchBar.waitForExistence(timeout: 1) {
            searchBar.tap()
            searchBar.typeText(query)
            sleep(1)
        }
        return self
    }

    @discardableResult
    func tapFilter() -> OrderHistoryPage {
        filterButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func filterByDateRange(_ range: DateRange) -> OrderHistoryPage {
        tapFilter()

        switch range {
        case .last7Days:
            if last7DaysButton.waitForExistence(timeout: 1) {
                last7DaysButton.tap()
            }
        case .last30Days:
            if last30DaysButton.waitForExistence(timeout: 1) {
                last30DaysButton.tap()
            }
        case .last3Months:
            if last3MonthsButton.waitForExistence(timeout: 1) {
                last3MonthsButton.tap()
            }
        }

        sleep(1)
        return self
    }

    @discardableResult
    func pullToRefresh() -> OrderHistoryPage {
        let scrollView = app.scrollViews.firstMatch
        if scrollView.exists {
            let start = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.2))
            let end = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.8))
            start.press(forDuration: 0, thenDragTo: end)
            sleep(1)
        }
        return self
    }

    @discardableResult
    func tapHelp() -> OrderHistoryPage {
        if helpButton.waitForExistence(timeout: 1) {
            helpButton.tap()
            sleep(1)
        }
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> OrderHistoryPage {
        XCTAssertTrue(firstOrder.waitForExistence(timeout: timeout) ||
                     emptyStateMessage.waitForExistence(timeout: timeout) ||
                     activeTab.waitForExistence(timeout: timeout),
                     "Order history screen should be displayed")
        return self
    }

    @discardableResult
    func assertOrdersDisplayed() -> OrderHistoryPage {
        XCTAssertTrue(firstOrder.exists, "Orders should be displayed")
        return self
    }

    @discardableResult
    func assertEmptyState() -> OrderHistoryPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    @discardableResult
    func assertOrderCount(_ expectedCount: Int) -> OrderHistoryPage {
        let actualCount = ordersList.count
        XCTAssertEqual(actualCount, expectedCount,
                      "Should have \(expectedCount) orders, found \(actualCount)")
        return self
    }

    @discardableResult
    func assertTabsDisplayed() -> OrderHistoryPage {
        XCTAssertTrue(activeTab.exists || completedTab.exists,
                     "Order tabs should be displayed")
        return self
    }

    @discardableResult
    func assertReorderButtonVisible() -> OrderHistoryPage {
        XCTAssertTrue(reorderButtons.firstMatch.exists,
                     "Reorder button should be visible")
        return self
    }

    @discardableResult
    func assertTrackButtonVisible() -> OrderHistoryPage {
        XCTAssertTrue(trackButtons.firstMatch.exists,
                     "Track button should be visible")
        return self
    }

    @discardableResult
    func assertOrderDetailsDisplayed() -> OrderHistoryPage {
        XCTAssertTrue(orderIdLabels.count > 0 ||
                     orderDateLabels.count > 0 ||
                     orderTotalLabels.count > 0,
                     "Order details should be displayed")
        return self
    }

    // MARK: - Types

    enum DateRange {
        case last7Days
        case last30Days
        case last3Months
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/OrderHistoryPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Order History Screen
 * Provides fluent API for order history interactions
 */
class OrderHistoryPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val ordersList = composeTestRule.onAllNodesWithTag("OrderItem")
    private val emptyStateMessage = composeTestRule.onNodeWithText("no order", substring = true, ignoreCase = true)
    private val activeTab = composeTestRule.onNodeWithText("active", substring = true, ignoreCase = true)
    private val completedTab = composeTestRule.onNodeWithText("completed", substring = true, ignoreCase = true)
    private val cancelledTab = composeTestRule.onNodeWithText("cancelled", substring = true, ignoreCase = true)
    private val filterButton = composeTestRule.onNodeWithText("filter", substring = true, ignoreCase = true)
    private val searchBar = composeTestRule.onNodeWithTag("search-bar")
    private val helpButton = composeTestRule.onNodeWithText("help", substring = true, ignoreCase = true)

    // Actions
    fun tapFirstOrder(): OrderHistoryPage {
        ordersList[0].performClick()
        Thread.sleep(1000)
        return this
    }

    fun tapOrder(index: Int): OrderHistoryPage {
        if (ordersList.fetchSemanticsNodes().size > index) {
            ordersList[index].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun switchToActiveOrders(): OrderHistoryPage {
        activeTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToCompletedOrders(): OrderHistoryPage {
        completedTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun switchToCancelledOrders(): OrderHistoryPage {
        cancelledTab.performClick()
        Thread.sleep(1000)
        return this
    }

    fun searchOrders(query: String): OrderHistoryPage {
        searchBar.performTextInput(query)
        Thread.sleep(1000)
        return this
    }

    fun tapFilter(): OrderHistoryPage {
        filterButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun filterByDateRange(range: DateRange): OrderHistoryPage {
        tapFilter()
        when (range) {
            DateRange.LAST_7_DAYS -> composeTestRule.onNodeWithText("Last 7 Days").performClick()
            DateRange.LAST_30_DAYS -> composeTestRule.onNodeWithText("Last 30 Days").performClick()
            DateRange.LAST_3_MONTHS -> composeTestRule.onNodeWithText("Last 3 Months").performClick()
        }
        Thread.sleep(1000)
        return this
    }

    fun reorderFirst(): OrderHistoryPage {
        composeTestRule.onNode(
            hasText("reorder", substring = true, ignoreCase = true)
        ).performClick()
        Thread.sleep(1000)
        return this
    }

    fun trackFirstOrder(): OrderHistoryPage {
        composeTestRule.onNode(
            hasText("track", substring = true, ignoreCase = true)
        ).performClick()
        Thread.sleep(1000)
        return this
    }

    fun tapHelp(): OrderHistoryPage {
        helpButton.performClick()
        Thread.sleep(1000)
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): OrderHistoryPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            ordersList.fetchSemanticsNodes().isNotEmpty() ||
            try { emptyStateMessage.assertExists(); true } catch (e: AssertionError) { false } ||
            try { activeTab.assertExists(); true } catch (e: AssertionError) { false }
        }
        return this
    }

    fun assertOrdersDisplayed(): OrderHistoryPage {
        assert(ordersList.fetchSemanticsNodes().isNotEmpty()) {
            "Orders should be displayed"
        }
        return this
    }

    fun assertEmptyState(): OrderHistoryPage {
        emptyStateMessage.assertExists()
        return this
    }

    fun assertOrderCount(expectedCount: Int): OrderHistoryPage {
        val actualCount = ordersList.fetchSemanticsNodes().size
        assert(actualCount == expectedCount) {
            "Should have $expectedCount orders, found $actualCount"
        }
        return this
    }

    fun assertTabsDisplayed(): OrderHistoryPage {
        activeTab.assertExists()
        return this
    }

    fun assertReorderButtonVisible(): OrderHistoryPage {
        composeTestRule.onNode(
            hasText("reorder", substring = true, ignoreCase = true)
        ).assertExists()
        return this
    }

    fun assertTrackButtonVisible(): OrderHistoryPage {
        composeTestRule.onNode(
            hasText("track", substring = true, ignoreCase = true)
        ).assertExists()
        return this
    }

    fun assertOrderDetailsDisplayed(): OrderHistoryPage {
        // Check if order details like ID, date, or total are displayed
        val hasOrderDetails = try {
            composeTestRule.onNode(hasText("₹", substring = true)).assertExists()
            true
        } catch (e: AssertionError) {
            false
        }
        assert(hasOrderDetails) {
            "Order details should be displayed"
        }
        return this
    }

    enum class DateRange {
        LAST_7_DAYS,
        LAST_30_DAYS,
        LAST_3_MONTHS
    }
}
